marine.c

```diff
@@ -579,6 +579,35 @@ WS_DLL_PUBLIC marine_result *marine_dissect_packet(int filter_id, unsigned char
     return result;
 }
 
+
+WS_DLL_PUBLIC int compile_bpf(char *bpf, struct bpf_program *fcode, char *err_msg) {
+    pcap_t *pc;
+
+    pc = pcap_open_dead(DLT_EN10MB, MIN_PACKET_SIZE);
+    if (pc != NULL) {
+        struct bpf_program temp_fcode;
+        struct bpf_program *used_fcode;
+        int ret_value;
+
+        used_fcode = (fcode != NULL) ? fcode : &temp_fcode;
+        ret_value = pcap_compile(pc, used_fcode, bpf, 0, 0);
+
+        if (ret_value == -1) {
+             strcpy(err_msg, "Failed compiling the BPF");
```

Comment:
Looks like the only reason you're not `return`-ing right here is because you don't want to duplicate the resource handling at the end (`pcap_close`).
Am I correct?
That's a good instinct but it makes your flow a bit convoluted. Try thinking of a better way